File: algorithms/260_SingleNumberIII.java
class Solution {
    public int[] singleNumber(int[] nums) {
        int res = 0;
        for (int num: nums) {
            res ^= num;
        }
        
        int i = 1;
        while ((res & i) == 0) {
            i <<= 1;
        }
        int num1 = 0, num2 = 0;
        for (int num: nums) {
            if ((num & i) != 0) {
                num1 ^= num;
            } else {
                num2 ^= num;
            }
        }
        return new int[]{num1, num2};
    }
}

File: algorithms/260_SingleNumberIII.py
class Solution:
    def singleNumber(self, nums):
        res = 0
        for num in nums:
            res ^= num
        i = 1
        while not res & i:
            i <<= 1
            
        num1, num2 = 0, 0
        for num in nums:
            if num & i:
                num1 ^= num
            else:
                num2 ^= num
        return [num1, num2]